HeatStroke/Graphics/Particles/Emitter.cpp

```diff
@@ -0,0 +1,702 @@
+	std::string strAffectorFunction = "";
+	EasyXML::GetRequiredStringAttribute(p_pAffectorFadeElement, "function", strAffectorFunction);
+	const char* cAffectorFunction = strAffectorFunction.c_str();
+	if (strcmp(cAffectorFunction, "lifetime") == 0)
+	{
+		eAffectorFunction = Affector::AFFECTOR_FUNCTION::AF_LIFETIME;
+	}
+
+	float fFade = 0.0f;
+	EasyXML::GetRequiredFloatAttribute(p_pAffectorFadeElement, "value", fFade);
+	m_vAffectors.push_back(new AffectorFade(eAffectorFunction, fFade));
+}
+
+void HeatStroke::Emitter::InitAffectorScale(tinyxml2::XMLElement* p_pAffectorScaleElement)
+{
+#if _DEBUG
+	assert(p_pAffectorScaleElement != nullptr);
+#endif
+
+	Affector::AFFECTOR_FUNCTION eAffectorFunction = Affector::AFFECTOR_FUNCTION::AF_CONSTANT;
+	std::string strAffectorFunction = "";
+	EasyXML::GetRequiredStringAttribute(p_pAffectorScaleElement, "function", strAffectorFunction);
+	const char* cAffectorFunction = strAffectorFunction.c_str();
+	if (strcmp(cAffectorFunction, "lifetime") == 0)
+	{
+		eAffectorFunction = Affector::AFFECTOR_FUNCTION::AF_LIFETIME;
+	}
+
+	float fScale = 0.0f;
+	EasyXML::GetRequiredFloatAttribute(p_pAffectorScaleElement, "amount", fScale);
+	m_vAffectors.push_back(new AffectorScale(eAffectorFunction, fScale));
+}
+
+void HeatStroke::Emitter::InitAffectorVelocity()
+{
+	m_vAffectors.push_back(new AffectorVelocity());
+}
+
+void HeatStroke::Emitter::InitParticles()
+{
+	Particle* pParticles = new Particle[m_iNumParticles];
+	for (int i = 0; i < m_iNumParticles; ++i)
+	{
+		AddParticleToPool(&pParticles[i]);
+	}
+}
+
+void HeatStroke::Emitter::AddParticleToPool(Particle* p_pParticle)
+{
+	p_pParticle->m_pPrev = nullptr;
+	p_pParticle->m_pNext = m_pFreeList;
+	m_pFreeList = p_pParticle;
+}
+
+void HeatStroke::Emitter::SpawnNewParticles(const float p_fDelta)
+{
+	int iNumParticlesToSpawn = 0;
+
+	if (m_bContinuous)
+	{
+		m_fDuration -= p_fDelta;
+		if (m_fDuration <= 0.0f && m_fDuration > -1.0f)
+		{
+			Stop();
+			return;
+		}
+
+		m_fToSpawnAcc += m_fBirthRate * p_fDelta;
+		iNumParticlesToSpawn = (int)m_fToSpawnAcc;
+	}
+	else
+	{
+		m_fTimeToBurst -= p_fDelta;
+		if (m_fTimeToBurst <= 0.0f)
+		{
+			iNumParticlesToSpawn = m_iNumParticles;
+
+			if (m_bRepeatBursts)
+			{
+				GetTimeToNextBurst();
+			}
+			else
+			{
+				Stop();
+			}
+		}
+	}
+
+	for (int i = 0; i < iNumParticlesToSpawn; ++i)
+	{
+		SpawnParticle();
+	}
+	m_fToSpawnAcc -= iNumParticlesToSpawn;
+}
+
+void HeatStroke::Emitter::UpdateActiveParticles(const float p_fDelta)
+{
+	Particle* pParticle = m_pActiveHead;
+	while (pParticle)
+	{
+		pParticle->m_fAge += p_fDelta * pParticle->m_fAgeFactor;
+		if (pParticle->m_fAge >= 1.0f)
+		{
+			Particle* pOldParticle = pParticle;
+			pParticle = pParticle->m_pNext;
+			KillParticle(pOldParticle);
+		}
+		else
+		{
+			auto it = m_vAffectors.begin();
+			auto end = m_vAffectors.end();
+			for (; it != end; ++it)
+			{
+				(*it)->Apply(p_fDelta, pParticle);
+			}
+			pParticle = pParticle->m_pNext;
```

Comment:
It would be, you're right.  For now it's not necessary, but it is an obvious optimization, so we'd might as well.  Will change it when making code less AIDS-y